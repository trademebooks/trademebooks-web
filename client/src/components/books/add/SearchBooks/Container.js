import { connect } from 'react-redux';
import { addBook } from '../../../../actions/book';
import SearchBooks from './SearchBooks';

const mapStateToProps = (state) => ({
  book: state.book.book
});

const mapDispatchToProps = {
  addBook
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(SearchBooks);
