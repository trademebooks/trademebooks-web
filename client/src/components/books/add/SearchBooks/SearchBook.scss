.search-books-container {
  border: solid 2px red;
  margin: 1.5rem 0 1.5rem 0;
}