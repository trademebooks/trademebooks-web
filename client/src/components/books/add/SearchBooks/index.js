import Container from './Container'; 

const SearchBooks = Container;

export default SearchBooks;

