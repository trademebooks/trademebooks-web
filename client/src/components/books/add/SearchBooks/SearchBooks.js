import React, { useState, useEffect } from 'react';
import TextField from '@material-ui/core/TextField';
import Autocomplete from '@material-ui/lab/Autocomplete';
import CircularProgress from '@material-ui/core/CircularProgress';

import './SearchBook.scss';

// async example: https://material-ui.com/components/autocomplete/#asynchronous-requests
const SearchBooks = ({ addBook, book }) => {
  const [value, setValue] = useState(null);
  const [inputValue, setInputValue] = useState('');

  const [open, setOpen] = useState(false);
  const [options, setOptions] = useState([]);
  const loading = open && options.length === 0;

  useEffect(() => {
    let active = true;

    if (!loading) {
      return undefined;
    }

    (async () => {
      if (inputValue.length > 2) {
        let url = `https://www.googleapis.com/books/v1/volumes?key=AIzaSyCpl497dKbKN-piJBJJ5zOf3sCPk7CKuJg&q=${inputValue}`;
        const response = await fetch(url);
        const data = await response.json();
        let books = data.items.map((book) => {
          let bookVolumeInfo = book['volumeInfo'];
          return {
            title: bookVolumeInfo['title'],
            authors: bookVolumeInfo['authors'],
            publisher: bookVolumeInfo['publisher'],
            description: bookVolumeInfo['description'],
            // image_url: bookVolumeInfo["imageLinks"]['smallThumbnail'],

            //image_url: bookVolumeInfo["imageLinks"]['smallThumbnail'],
            //isbn_10: bookVolumeInfo['industryIdentifiers'][0]['identifier'],
            // isbn_13: bookVolumeInfo['industryIdentifiers'][1]['identifier']
          };
        });

        //console.log({ books });

        if (active) {
          setOptions(books);
        }
      }
    })();

    return () => {
      active = false;
    };
  }, [inputValue]);

  useEffect(() => {
    if (!open) {
      setOptions([]);
    }
  }, [open]);

  return (
    <>
      <div className="search-books-container">
        <div className="section-header">
          <p>SEARCH FROM OVER 10,000 BOOKS</p>
        </div>
        <div className="search-box">
          <Autocomplete
            id="asynchronous-demo"
            style={{ width: '100%' }}
            open={open}
            onOpen={() => {
              setOpen(true);
            }}
            onClose={() => {
              setOpen(false);
            }}
            getOptionSelected={(option, value) => option.title === value.title}
            getOptionLabel={(option) => option.title}
            options={options}
            loading={loading}
            value={value}
            onChange={(event, newValue) => {
              setOptions(newValue ? [newValue, ...options] : options);
              setValue(newValue);
              addBook(newValue);
            }}
            onInputChange={(event, newInputValue) => {
              setInputValue(newInputValue);
            }}
            renderInput={(params) => (
              <TextField
                {...params}
                label="Type in a book name, author name, or isbn number..."
                variant="outlined"
                InputProps={{
                  ...params.InputProps,
                  endAdornment: (
                    <>
                      {loading ? (
                        <CircularProgress color="inherit" size={20} />
                      ) : null}
                      {params.InputProps.endAdornment}
                    </>
                  ),
                }}
              />
            )}
            renderOption={(option) => {
              if (option.book_image && option.book_image.smallThumbnail) {
                return (
                  <div>
                    <span>
                      <img src={option.book_image.smallThumbnail} alt={option.title}/>
                    </span>{' '}
                    {option.title}
                  </div>
                );
              } else {
                return (
                  <div>
                    <span></span> {option.title}
                  </div>
                );
              }
            }}
          />
        </div>
        <div className="create-new-listing">
          <p>
            Can't find your book? <a href="/">Create Listing Manually</a>
          </p>
        </div>
      </div>
    </>
  );
};

export default SearchBooks;
