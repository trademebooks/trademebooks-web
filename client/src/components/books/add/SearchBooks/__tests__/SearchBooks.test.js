import React from 'react'
import renderer from 'react-test-renderer'

import SearchBooks from '../SearchBooks'

describe('Component: SearchBooks', () => {
  const props = {
    addBook: jest.fn()
  }

  it('should match the snapshot', () => {
    const SearchBooksSnapshot = renderer
      .create(<SearchBooks {...props} />)
      .toJSON()

    expect(SearchBooksSnapshot).toMatchSnapshot()
  })
})
