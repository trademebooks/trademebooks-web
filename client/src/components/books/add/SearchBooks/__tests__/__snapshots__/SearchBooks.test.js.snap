// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: SearchBooks should match the snapshot 1`] = `
<div
  className="search-books-container"
>
  <div
    className="m-2 text-center"
    id="search-books-header"
  >
    <p
      className="search-books-header"
    >
      Search over millions of books!
    </p>
  </div>
  <div>
    <div
      aria-expanded={false}
      aria-owns={null}
      className="MuiAutocomplete-root MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon"
      onClick={[Function]}
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      role="combobox"
    >
      <div
        className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink={false}
          htmlFor="autocomplete-box"
          id="autocomplete-box-label"
        >
          Type in a book name, author name, or isbn number...
        </label>
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
          onClick={[Function]}
        >
          <input
            aria-activedescendant={null}
            aria-autocomplete="list"
            aria-controls={null}
            aria-invalid={false}
            autoCapitalize="none"
            autoComplete="off"
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
            disabled={false}
            id="autocomplete-box"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onMouseDown={[Function]}
            required={false}
            spellCheck="false"
            type="text"
            value=""
          />
          <div
            className="MuiAutocomplete-endAdornment"
          >
            <button
              aria-label="Clear"
              className="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-clearIndicator"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={-1}
              title="Clear"
              type="button"
            >
              <span
                className="MuiIconButton-label"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </svg>
              </span>
            </button>
            <button
              aria-label="Open"
              className="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-popupIndicator"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={-1}
              title="Open"
              type="button"
            >
              <span
                className="MuiIconButton-label"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
              </span>
            </button>
          </div>
          <fieldset
            aria-hidden={true}
            className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="PrivateNotchedOutline-legendLabelled-3"
            >
              <span>
                Type in a book name, author name, or isbn number...
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
  </div>
</div>
`;
